src/edata_cache.c

```diff
@@ -0,0 +1,47 @@
+#include "jemalloc/internal/jemalloc_preamble.h"
+#include "jemalloc/internal/jemalloc_internal_includes.h"
+
+bool
+edata_cache_init(edata_cache_t *edata_cache) {
+	if (malloc_mutex_init(&edata_cache->mtx, "edata_cache",
+	    WITNESS_RANK_EDATA_CACHE, malloc_mutex_rank_exclusive)) {
+		return true;
+	}
+
+	edata_avail_new(&edata_cache->avail);
+	return false;
+}
+
+edata_t *
+edata_cache_alloc(tsdn_t *tsdn, edata_cache_t *edata_cache, base_t *base) {
+	malloc_mutex_lock(tsdn, &edata_cache->mtx);
+	edata_t *edata = edata_avail_first(&edata_cache->avail);
+	if (edata == NULL) {
+		malloc_mutex_unlock(tsdn, &edata_cache->mtx);
+		return base_alloc_edata(tsdn, base);
+	}
+	edata_avail_remove(&edata_cache->avail, edata);
+	atomic_fetch_sub_zu(&edata_cache->count, 1, ATOMIC_RELAXED);
+	malloc_mutex_unlock(tsdn, &edata_cache->mtx);
+	return edata;
+}
+
+void
+edata_cache_dalloc(tsdn_t *tsdn, edata_cache_t *edata_cache, edata_t *edata) {
```

Comment:
nit: I kind of want a name to indicate this is returning to the cache (dealloc can be a bit ambiguous), if you can find one (I tried but couldn't). Feel free to ignore.

Reply:
acquire/release maybe? Or remove/insert?

I also like not naming this alloc/dalloc just because edata_cache_alloc and ecache_alloc is a recipe for confusion.

Reply:
I slightly lean towards remove/insert.

Reply:
Talked offline: get/put